code/modules/hydroponics/grown/towercap.dm

```diff
@@ -179,6 +182,48 @@
 	return ..()
 
 /obj/structure/bonfire/attackby(obj/item/W, mob/user, params)
+	//SMOKE SIGNALS. Burn cloth to send a radio message to everyone who knows tribal.
+	if(istype(W, /obj/item/stack/sheet/cloth) && !grill && burning)
+		var/obj/item/stack/sheet/cloth/C = W
+		if(!user.has_language(/datum/language/tribal))
+			return ..()
+		var/outdoors = FALSE
+		for(var/area_type in OUTDOOR_AREAS)
+			if(istype(get_area(src), area_type))
+				outdoors = TRUE
+				break
+		if(!outdoors)
+			to_chat(user, span_warning("You must be outside to send a smoke signal."))
+			return
+		var/signalmessage = stripped_input(user, "What would you like to send via smoke signal?", "Smoke Signal")
+		if(!signalmessage)
+			return
+		var/send_time = length(signalmessage) * 2 //Deciseconds. A 10 character message would take 20 deciseconds, or two seconds, to send.
+		if(!user.canUseTopic(src, BE_CLOSE))
+			to_chat(user, span_warning("You have to stand still to send a smoke signal."))
```

Comment:
After we have a valid signalmessage, I'd prefer if the failure messages also included that message, so its not lost and players can just copypaste it back in.

Reply:
That's a very good point. I wish there was a way to make it automatically fill in the last attempted message if it failed, but that would require setting a variable _somewhere_, and I don't think the player, the cloth, _or_ the bonfire really needs a variable like that.

Reply:
We already have the message stored in `signalmessage`, would be enough to just do something like:

`to_chat(user, span_warning("You don't have enough cloth to send a smoke signal. original message: [signalmessage]"))`